Cookiebanner + Umami tracking

JulianNymark opened the pull request.

JulianNymark committed:
:construction: umami test script

JulianNymark committed:
:construction: CSP for umami

JulianNymark committed:
:construction: test data-tag, set no auto tracking

JulianNymark committed:
Merge branch 'main' into umami

JulianNymark committed:
:recycle: make consentBanner more self contained

JulianNymark committed:
:construction: working implementation for consent modal

JulianNymark committed:
Merge branch 'cookiebanner' into next

JulianNymark committed:
:recycle: move ConsentBanner to website-modules + small changes

JulianNymark committed:
use typescript-cookies over localStorage API

JulianNymark committed:
Copy Nav.no consent text

KenAJoh commented on aksel.nav.no/website/components/website-modules/ConsentBanner.tsx:
```diff
@@ -0,0 +1,137 @@
+      ref.current?.showModal();
+    }
+
+    const isProdUrl = () => window.location.host === "aksel.nav.no";
```
Think it would make sense to extract this to a util function `getCurrentEnviroment`or something. Think we should account for if `example` or `template` is viewed to, so
```
const isProdUrl = () => window.location.host === "aksel.nav.no";
    const isExample = () => window.location.pathname.includes("eksempler/");
    const isTemplate = () => window.location.pathname.includes("templates/");
    const isPreview = () => !!document.getElementById("exit-preview-id");
```

Might also make sense to turn of tracking when on example/template 🤔 If you visit a component-page, we dont want to show that as extra view when the example loads (umami might account for this, so no idea if needed)

JulianNymark replied:
Yeah, i looked a bit at the logged data and it has some interesting things too, for example when using the admin panel / sanity studio it even logs that 😅 

![image](https://github.com/user-attachments/assets/d93251a2-7566-47a3-9eca-69b29d68872b)

JulianNymark replied:
a smallish problem with tags is that there seems to only be the possibility of having one tag at a time?

But i can easily see something like `production` and `eksempler/` apply at the same time 🤔. So maybe we only want to have environments for tags, and then filter on path query later? (I guess all of these could probably be somewhat easily deciphered from the URL of the log entry, so maybe we should just remove the tags entirely?)

JulianNymark replied:
I guess it's another point in favor of tracking manually curated objects 
![image](https://github.com/user-attachments/assets/7a962652-a796-4238-b784-3f47340a5f54)

JulianNymark replied:
Think I've got a good solution now for using the auto tracker + setting a useful tag.

Landing on two tags: `organic` (for natural organic traffic we care about), and `polluted` (which might still be useful, but require a bit more effort to filter out later, usually I think we can do a very simple filter on a path matcher `/admin/` for example, to see how much traffic there is in sanity studio or similar.)

KenAJoh commented on aksel.nav.no/website/components/website-modules/ConsentBanner.tsx:
```diff
@@ -0,0 +1,106 @@
+
+  return (
+    <div>
+      {refUmamiTag.current && (
```
This will never update since its a ref

JulianNymark replied:
I made it a ref (had state first, but then encountered first render issues when I didn't want the umami script blob to run _before_ it had the latest value, where the tag is not getting set since the tag needs to run some js logic first to determine what tag to set 🤔 )

I can probably find a way to make this work without using a ref, maybe via a child component or similar.

JulianNymark replied:
I was also thinking that it doesn't _need_ to update? people are unlikely to update the tag within a single tab? (well... maybe we are when loading the admin panel... not entirely sure)

KenAJoh replied:
Ment more of that updating this snippet 
```
refUmamiTag.current = isPreview()
      ? "preview"
      : isProdUrl()
        ? "production"
        : "development";
```
Does not cause `refUmamiTag.current &&` to update. It only works now since `clientAcceptsTracking`-state is updated, so only as a side-effect.

JulianNymark replied:
I guess I wanted a value to be immediately available within _this_ node (for consumption in the render function), but I didn't care for it triggering any re-renders if it changed (since it's not really likely to change after first page visit?).

If i used `useState()` the change would only exist on the next render pass after calling the `setState...()`, and by then it's too late, the JS inside the <Script> has run, and set the tag for sending tracking events 🤔... I can demo this too.

KenAJoh commented on aksel.nav.no/website/components/website-modules/ConsentBanner.tsx:
```diff
@@ -0,0 +1,106 @@
+          data-tag={refUmamiTag.current}
+        ></Script>
+      )}
+      <Modal
```
We give the user no option to "refuse to answer" now. The user should be able to exit out without answering. If they do that, we can either assume they want no tracking, or we create a temporary cookie that expires after they stop the session. If that temp-cookie is set, we assume the user wants no tracking and wont bug them until the next time they visit.

JulianNymark replied:
You can click the "X", and it doesn't remember this, but it's effectively going to last for that tabs "session", since the node is high up in the tree, and it's a `useEffect(..., [])` 🤔. 

My first version had explicit docs on what options you are allowed to take "press X" vs "press Yes" vs "press No", and what happens when you do.... a bit verbose though.

JulianNymark replied:
So I guess making a very short term cookie to remember "no action" is what we should add 🤔 Makes sense!

KenAJoh commented on aksel.nav.no/website/components/website-modules/ConsentBanner.tsx:
```diff
@@ -0,0 +1,106 @@
+
+const CONSENT_TRACKER_ID = "acceptTracking";
+
+type CONSENT_TRACKER_STATE = "undecided" | "accepted" | "rejected";
```
Think we can take some inspiration from dekoratoren and have a more future proof state. Something like this
```
{
version: 1,
optional: {
  umami: null // true/false/?
}
```
This allows us to "invalidate" the consent of version updates, or we add new optional consents.

KenAJoh commented on aksel.nav.no/website/components/website-modules/ConsentBanner.tsx:
```diff
@@ -0,0 +1,106 @@
+};
+
+export const setStorageAcceptedTracking = (state: CONSENT_TRACKER_STATE) => {
+  Cookies.set(CONSENT_TRACKER_ID, state, { expires: 365 });
```
We should set `{ domain: 'aksel.nav.no' }` in the options so that the cookie only shows up when visiting Aksel. 

Unsure how this will affect testing on `aksel.ansatt.nav.no` tho

JulianNymark replied:
since the code is only served via our domains, do we need to limit the cookie by domain specifically? (I guess iframes linking to us could get the cookie banner?)

it-vegard replied:
Setting `{ domain: location.host }` would probably allow for separate cookie for testing on aksel.ansatt.nav.no, which I guess is what we want?

JulianNymark replied:
this tracking ID was generated to apply for aksel.ansatt.dev currently (if you see it in the Umami web-UI), we can get a new tracking ID for prod from Eilif and switch it out based on current domain 🤔 .

KenAJoh commented on aksel.nav.no/website/components/website-modules/ConsentBanner.tsx:
```diff
@@ -0,0 +1,106 @@
+import { Cookies } from "typescript-cookie";
+import { BodyLong, Button, Link, Modal } from "@navikt/ds-react";
+
+const CONSENT_TRACKER_ID = "acceptTracking";
```
To make it more distinguishable and generic something like `aksel-consent` could be used as name.

JulianNymark replied:
I was thinking here that it's already a _per domain_ type of data (to see it in the browser application/storage tabs, it shows you the domain at the top level)
![image](https://github.com/user-attachments/assets/0ef9f2d0-5ca4-449f-8fc0-f6176dcdd903)

But no big prefs either

KenAJoh replied:
Here i though more of just having the prefix `aksel` so when i scan over all the possible tokens in "application" i can quickly find "aksel-" in between all other names

JulianNymark committed:
better name for cookie

JulianNymark committed:
:construction: crude complex consent object

KenAJoh commented on aksel.nav.no/website/components/website-modules/ConsentBanner.tsx:
```diff
@@ -0,0 +1,126 @@
+  updatedAt: string;
+  version: number;
+  consents: {
+    tracking?: string;
```
```suggestion
    tracking?: CONSENT_TRACKER_STATE;
```

JulianNymark commented on aksel.nav.no/website/components/website-modules/ConsentBanner.tsx:
```diff
@@ -0,0 +1,126 @@
+            er sikkert, og kan ikke velges bort. Andre brukes til statistikk og
+            analyse. Godkjenner du alle, hjelper du oss å lage bedre nettsider
+            og tjenester.{" "}
+            <Link href="https://www.nav.no/informasjonskapsler">
```
link to our own

KenAJoh commented on aksel.nav.no/website/components/website-modules/ConsentBanner.tsx:
```diff
@@ -0,0 +1,126 @@
+          src="https://cdn.nav.no/team-researchops/sporing/sporing.js"
+          data-host-url="https://umami.nav.no"
+          data-website-id="7b9fb2cd-40f4-4a30-b208-5b4dba026b57"
+          data-auto-track={clientAcceptsTracking ? "true" : "false"}
```
```suggestion
          data-auto-track={clientAcceptsTracking}
```

JulianNymark committed:
Update aksel.nav.no/website/components/website-modules/ConsentBanner.tsx

Co-authored-by: Ken <26967723+KenAJoh@users.noreply.github.com>

JulianNymark committed:
classify traffic better (umami)

JulianNymark committed:
Update aksel.nav.no/website/components/website-modules/ConsentBanner.tsx

Co-authored-by: Ken <26967723+KenAJoh@users.noreply.github.com>

JulianNymark committed:
link to our own personvernerklering + disable modal on that specific page

JulianNymark committed:
use same conditions for all showModal() triggers

JulianNymark commented on aksel.nav.no/website/components/website-modules/ConsentBanner.tsx:
```diff
@@ -0,0 +1,143 @@
+    if (consentAnswer === "undecided" && !disabledModalParam) {
+      ref.current?.showModal();
+    }
+    let previousPath = "";
```
hold on, lemme explain!

first i thought it makes sense to exclude the page that explains our privacy policy from having the consent banner modal presented, but then i thought about how that shouldn't permanently pick on your behalf either, so the banner should "come back" once you're done reading more info _about_ the banner (and any other privacy policy stuff), the modal makes it somewhat hard to do that, so it _should_ be excluded on that page, and not have you make a choice before you get the info you might need in order to _make_ said choice.

Our website is an SPA, so the best i could do to "ad hoc" detect location changes was a MutationObserver on the entire document and check the location.

It needs some cleanup too, hence the disconnect().

The logic also only checks previousPath and if that matches the currently hardcoded path to the privacy policy (not great 😬 ).

Additionally the privacy policy page has links going to it that set the query parameter `no_consent_modal` which prevents the banner from loading on first load! (which could happen... maybe you link the PP to someone else).

Lastly the link to the PP from the banner _itself_ has a modal.close(), as it'll be re-shown later when applicable (navigate away from PP or reload).

tl;dr:
- Shouldn't show on first load (if you have the query param set)
- Goes away when you click the link to more info in banner.
- Comes back when you leave info on banner (or reload the page, but not if you reload the page on the info about the banner)

This logic exploded 💥 in complexity 😢

JulianNymark replied:
Is this sane/insane? I'm ok with throwing all of this away for simplicity, even if the solution is less feature packed.

IMO it's not the end of the world to show a modal on the privacy policy page... just going to assume that most people don't go there 😅

KenAJoh replied:
Leaning towards insane here i think 😆 I think a non-modal approach would make this a non-issue since it won't be as intrusive.

JulianNymark committed:
update comment

JulianNymark committed:
limit cookie to specific subdomain (prevent it getting sent by clients to other *.nav.no pages

JulianNymark committed:
remove the MutationObserver, but keep query param to disable banner (for tests)

JulianNymark committed:
bypass banner in e2e tests

JulianNymark committed:
store "no action" consent data in sessionStorage

JulianNymark committed:
Merge branch 'main' into next

JulianNymark committed:
Merge branch 'main' into next

JulianNymark commented on aksel.nav.no/website/components/website-modules/ConsentBanner.tsx:
```diff
@@ -0,0 +1,162 @@
+ * therefore we have to check if the user closed via one of the buttons
+ * (a cookie or sessionStorage was set) or if it was simply closed via the 'X' (no action taken)
+ * this is where we can then act on "no action taken"*/
+// NOTE: maybe we should have a specific handler for "no action",
```
Is this a good idea? 🤔 (adding a specific handler function for the "x" ?)

JulianNymark replied:
no longer relevant to this PR, as we removed the modal, but might still be an idea for the modal user-facing API?

JulianNymark committed:
:construction: change to in document flow style banner (no more modal)

JulianNymark committed:
:iphone: responsive consent banner++

JulianNymark committed:
website: set cookie domain dynamically (from serving site)

JulianNymark committed:
Merge branch 'main' into next

JulianNymark committed:
hardcode /side/personvernerklaering

JulianNymark committed:
:construction: placeholder cookie preferences box

JulianNymark committed:
cookie preferences box++

JulianNymark committed:
dokmøte plenumsendringer

JulianNymark committed:
move page personvernserklaering

JulianNymark committed:
make the cookie preference form functional

JulianNymark committed:
:wheelchair: add missing aria-labelledby target

JulianNymark committed:
re-enable banner on e2e search tests

larseirikhansen committed:
Add explanation to organic and polluted traffic

larseirikhansen commented on aksel.nav.no/website/pages/personvernerklaering.tsx:
```diff
@@ -0,0 +1,198 @@
+                Opplysningene brukes til å kartlegge hvordan og hvor mye
+                aksel.nav.no brukes, uten å identifisere IP-adresser. Vi bruker
+                verktøyet Amplitude i analysearbeidet.
+              </BodyLong>
```
Umami!

> Vi bruker verktøyet **_Amplitude_** i analysearbeidet.

larseirikhansen commented on aksel.nav.no/website/pages/personvernerklaering.tsx:
```diff
@@ -0,0 +1,198 @@
+  } else if (data.get("acceptedTracking") === "tracking_no") {
+    setStorageAcceptedTracking("rejected");
+  }
+  return false;
```
Returverdien brukes ikke som jeg kan se

```suggestion
```

JulianNymark committed:
extract consent state into custom hook

JulianNymark committed:
PR review changes

KenAJoh committed:
:sparkles: Serverside-check for cookie consent

JulianNymark committed:
Merge branch 'main' into next

JulianNymark committed:
banner set tracking id from env var

JulianNymark committed:
Merge branch 'main' into next

KenAJoh commented on aksel.nav.no/website/components/website-modules/ConsentBanner.tsx:
```diff
@@ -0,0 +1,119 @@
+  const trackingId = getConfig().publicRuntimeConfig.UMAMI_TRACKING_ID;
+
+  useEffect(() => {
+    const disabledModalParam = new URLSearchParams(window.location.search).get(
```
`no_consent_modal` looks to be unused, so logic can be removed

KenAJoh commented on aksel.nav.no/website/next.config.js:
```diff
@@ -56,17 +56,21 @@ const nextConfig = {
   publicRuntimeConfig: {
     NEXT_PUBLIC_TEST: process.env.NEXT_PUBLIC_TEST,
+    UMAMI_TRACKING_ID:
+      process.env.NODE_ENV === "production"
```
Not 100% sure here, but as long as one runs "next build" this is set to production. This means that builds for "Aksel.ansatt.nav.no" also gets the production-tag. Is this intended?

JulianNymark replied:
correct, becomes development for `yarn dev` but production with `yarn build`, maybe not the best to use `NODE_ENV` for this, i'll look for a better existing value we could hopefully use to properly distinguish dev&prod for this before making something from scratch.... 🤔

JulianNymark committed:
remove unused url param check

JulianNymark committed:
env var toggle for PRODUCTION

use this to inject code scripts clientside, do stuff during build or anything else really! (better to use this than NODE_ENV)

JulianNymark committed:
Merge branch 'main' into next
